src/aiperf/credit/callback_handler.py

```
@@ -338,6 +414,26 @@ async def on_credit_return(
+        # gated strategy path. The gated path no-ops once the phase can no
+        # longer send (or when no handler is registered), which would strand
+        # the executor coroutine awaiting this node's dispatch. Firing here
+        # guarantees every graph credit's Future is resolved or rejected.
+        if credit.trace_id is not None and self._graph_return_observer is not None:
+            # Contain the observer: an exception here would skip counting, slot
+            # release, and the drain event, hanging the phase on its drain wait.
+            try:
+                self._graph_return_observer(
+                    credit, credit_return.error, credit_return.cancelled
+                )
+            except Exception as exc:
```

Ruff (0.16.0)

[warning] Do not catch blind exception: `Exception`

(BLE001)